addons/point_of_sale/models/pos_session.py

```diff
@@ -806,25 +786,33 @@ def _accumulate_amounts(self, data):
                 partners._increase_rank('customer_rank')
 
         if self.company_id.anglo_saxon_accounting:
-            global_session_pickings = self.picking_ids.filtered(lambda p: not p.pos_order_id)
-            if global_session_pickings:
-                stock_moves = self.env['stock.move'].sudo().search([
-                    ('picking_id', 'in', global_session_pickings.ids),
+            all_picking_ids = self.order_ids.filtered(lambda p: not p.is_invoiced).picking_ids.ids + self.picking_ids.filtered(lambda p: not p.pos_order_id).ids
+            if all_picking_ids:
+                # Combine stock lines
+                stock_move_sudo = self.env['stock.move'].sudo()
+                stock_moves = stock_move_sudo.search([
+                    ('picking_id', 'in', all_picking_ids),
                     ('company_id.anglo_saxon_accounting', '=', True),
                     ('product_id.categ_id.property_valuation', '=', 'real_time'),
                 ])
-                for move in stock_moves:
-                    exp_key = move.product_id._get_product_accounts()['expense']
-                    out_key = move.product_id.categ_id.property_stock_account_output_categ_id
-                    signed_product_qty = move.product_qty
-                    if move._is_in():
-                        signed_product_qty *= -1
-                    amount = signed_product_qty * move.product_id._compute_average_price(0, move.quantity_done, move)
-                    stock_expense[exp_key] = self._update_amounts(stock_expense[exp_key], {'amount': amount}, move.picking_id.date, force_company_currency=True)
-                    if move._is_in():
-                        stock_return[out_key] = self._update_amounts(stock_return[out_key], {'amount': amount}, move.picking_id.date, force_company_currency=True)
-                    else:
-                        stock_output[out_key] = self._update_amounts(stock_output[out_key], {'amount': amount}, move.picking_id.date, force_company_currency=True)
+
+                for stock_moves_split in self.env.cr.split_for_in_conditions(stock_moves.ids):
```

Comment:
@ryv-odoo Because I have to call `with_context`, and without splitting, I would have to call `with_context` 1000 times.

Reply:
I don't get it sorry. Now you make one `with_context` for each 1000 moves. If you remove this slit of ids, we will have only one `with_context` for all moves.

```python
              candidates = stock_moves\
                    .filtered(lambda m: not bool(m.origin_returned_move_id and sum(m.stock_valuation_layer_ids.mapped('quantity')) >= 0))\
                    .mapped('stock_valuation_layer_ids')
               for move in stock_moves.with_context(candidates_prefetch_ids=candidates._prefetch_ids):
```

I miss something ?

Reply:
@ryv-odoo I tested it, and it was 50% slower. I think it was slower because calling `with_prefetch` each time retrieves 100k records, and the overhead makes it slower.

Reply:
`with_prefetch` do nothing expected recreating a recordset with the new prefetchset. Moreover you don't call it more often. It certainly slower because the prefectset become bigger and because the ORM needs to traverse it after.